File: 1LL,stack,queue/d1_SLL.java
import java.util.Scanner;
class node{
    int value;
    node next;
}

public class d1_SLL {
    Scanner sc =new Scanner(System.in);
    node start;

    public static void main(String[] args) {
        d1_SLL list = new d1_SLL();
        d1_SLL list2 = new d1_SLL();
        list2.start= null;
       list.start = null;
       list.traverse();

       int choice;
       char ch;
       do{
            System.out.println("insert : \n1 for insertion_at_begin\n2 for insertion_at_end\n3 for insertion_at_position\n4 for delete_by_position\n5 for delete_by_value\n6 for search\n7 for reverse\n8 for ascending\n9 for clone");
            System.out.printf("enter choice : ");
            choice =  list.sc.nextInt();

            switch(choice) {
                case 1:
                    list.insertion_at_begin();
                    list.traverse();
                    break;

                case 2:
                    list.insertion_at_end();
                    list.traverse();
                    break;

                case 3:
                    list.insertion_at_position();
                    list.traverse();
                    break;

                case 4:
                    list.delete_by_position();
                    list.traverse();
                    break;

                case 5:
                    list.delete_by_value();
                    list.traverse();
                    break;

                case 6:
                    list.search();
                    list.traverse();
                    break;

                case 7:
                    list.reverse();
                    list.traverse();
                    break;

                case 8:
                    list.ascending();
                    list.traverse();
                    break;

                case 9:
                    list2.clone(list);
                    System.out.printf("cloned ");
                    list2.traverse();
                    System.out.printf("original ");
                    list.traverse();
                    break;

                default :
                System.exit(0);
            }
            System.out.println("\nDo you want to insert an another choice?");
            System.out.printf("Insert y for yes or n for no : ");
            ch  = list.sc.next().charAt(0);
        }while(ch =='y' || ch=='Y');
    }

    void traverse() {
        if(start == null) {
            System.out.println("empty linked list");
        }else {
            node temp = start;
            System.out.printf("list are as follows :");
            while(temp!=null) {
                System.out.printf(" " + temp.value);
                temp=temp.next;
            }
            System.out.println("\n");
        }
    }

    void insertion_at_begin() {
        node new1 =new node();
        System.out.printf("enter the value : ");
        new1.value=sc.nextInt();
        new1.next= start;
        start = new1;
    }

    void insertion_at_end() {
        node new1 = new node();
        System.out.printf("enter the value : ");
        new1.value=sc.nextInt();
        if(start == null) {
            new1.next=null;
            start = new1;
        }else{
            node temp=start;
            while(temp.next!=null) {
                temp= temp.next;
            }
            new1.next= temp.next;
            temp.next= new1;
        }
    }

    void insertion_at_position() {
        int count=0;
        node temp=start;
            while(temp!=null) {
                count++;
                temp=temp.next;
            }
        if(count==0) {
            System.out.println("no linked list");
        }else if(count==1){
            System.out.println("linked list should have more than 1 node");
        }else {
            int pos;
            System.out.printf("enter the position : ");
            pos=sc.nextInt();
            
            if(pos<2 || pos>count) {
                System.out.println("the position should be between 2 and " + count);
            }else{
                temp=start;
                for(int i=1;i<pos-1;i++) {
                    temp=temp.next;
                }
                node new1=new node();
                System.out.printf("enter the value : ");
                new1.value=sc.nextInt();
                new1.next=temp.next;
                temp.next=new1;
            }
        }
    }

    void delete_by_position() {
        if(start == null) {
            System.out.println("no linked list");
        }else if(start.next==null) {
            start = null;
            System.out.println("only node deleted");
        }else {
            int pos,count=0;
            node temp = start;
            while(temp!=null) {
                temp= temp.next;
                count++;
            }
            do{
                System.out.printf("enter the position only between 1 and " + count + " : ");
                pos = sc.nextInt();
            }while(pos<1 || pos>count);

            if(pos==1) {
                start = start.next;
                System.out.println("Node at position " + pos + " deleted.");
            }else {
                temp = start;
                for(int i=1;i<pos-1;i++) {
                    temp=temp.next;
                }
                temp.next=temp.next.next;
                System.out.println("Node at position " + pos + " deleted.");
            }

        }
    }

    void delete_by_value() {
        if(start==null) {
            System.out.println("no liked list");
        }else  {
            int val,pos=0,found=0;
            System.out.printf("enter the value : ");
            val = sc.nextInt();
            node temp = start;
            while(temp!=null) {
                pos++;
                if(val==temp.value) {
                    found=1;
                    break;
                }
                temp= temp.next;
            }
            if(found ==0) {
                System.out.println("value " + val+ " doesn't exist in list");
            }else {
                if(start.next==null) {
                    start = null;
                    temp=null;
                    System.out.println("node deleted of value  " + val);
                }else if(pos == 1) {
                    start= start.next;
                    temp=null;
                    System.out.println("node deleted of value  " + val);
                }else {
                    temp = start;
                    for(int i=1;i<pos-1;i++) {
                        temp=temp.next;
                    }
                    temp.next=temp.next.next;
                    System.out.println("node deleted of value  " + val);
                }
            }
        }
    }

    void search() {
        if(start==null) {
            System.out.println("no linked list");
        }else {
            int val,pos=0,found=0;
            System.out.printf("enter the value : ");
            val = sc.nextInt();
            node temp = start;
            while(temp!=null) {
                pos++;
                if(val==temp.value) {
                    found=1;
                    System.out.println("ys "+val+ " exist in list at position "+ pos);
                    break;
                }
                temp=temp.next;
            }
            if(found==0) {
                System.out.println("no " + val+ " doesn't exist in list");
            }
        }
    }

    void reverse() {
        if(start == null) {
            System.out.println("no linked list");
        }else {
            node prev=null;
            node curr= start;
            node nex = null;

            while(curr!=null) {
                nex=curr.next;
                curr.next=prev;
                prev=curr;
                curr=nex;
            }
            start=prev;

        }
    }

    void ascending() {
        if(start== null) {
            System.out.println("no linked list");
        }else{
            node temp = start;
            node temp1;
            int store;
            while(temp!=null) {
                temp1=  start;
                while(temp1!= null) {
                    if(temp.value<temp1.value) {
                        store = temp.value;
                        temp.value= temp1.value;
                        temp1.value= store;
                    }
                    temp1= temp1.next;
                }
                temp = temp.next;
            }
        }
    }

    void clone(d1_SLL list) {
        if(list.start == null) {
            System.out.println("no linked list");
        }else{
           node temp  = list.start;
           node temp1= start;
           while(temp!=null) {
                node new1 = new node();
                new1.value= temp.value;
                if(temp == list.start) {
                    new1.next = start;
                    start = new1;
                    temp1= start;
                }else{
                    new1.next = null;
                    temp1.next =  new1;
                    temp1= temp1.next;
                }
                temp =  temp.next;
            } 
        }
    }
   
}  

File: 1LL,stack,queue/d1_SLL.c
# include <stdio.h>
# include <stdlib.h>

struct link
{
	int info;
	struct link *next;
};

struct link *start, *new1;

void insertion_at_begin();
void insertion_at_end();
void insertion_at_position();
void delete_by_position();
void delete_by_value();
void search();
void reverse();
void traverse ();
void short_ascending();
void shivam_reverse();

/* Function main */
void main()
{
	int choice;
	char ch;

	start = NULL;      /* Empty list */
	traverse();

	do
	{
		printf("\n\nInsert\n1 for insert_at_beginning\n2 for insert_at_end\n3 for insert_at_position\n4 delete_by_position\n5 delete_by_value\n6 search\n7 reverse\n8 for short_ascending\n9 for shiva_reverse");

		printf("\n Insert choice: ");
		scanf("%d", &choice);
		getchar();

		switch(choice)
		{
			case 1:
				insertion_at_begin ();
				traverse ();
				break;

			case 2:
				insertion_at_end ();
				traverse ();
				break;

			case 3:
				insertion_at_position();
				traverse();
				break;

			case 4:
				delete_by_position();
				traverse();
				break;

			case 5:
				delete_by_value();
				traverse();
				break;

			case 6:
				search();
				traverse();
				break;

			case 7:
				reverse();
				traverse();
				break;

			case 8 :
				short_ascending();
				traverse();
				break;

			case 9:
				shivam_reverse();
				traverse();
				break;

			default:
				exit(1);
		}
		printf("\n\n Do you want to insert an another choice?  ");
		printf("\n Insert y for yes or n for no.  ");
		//fflush(stdin);
		scanf("%c",&ch);
	}while(ch == 'y' || ch == 'Y');
}

/* Function to insert a node at the beginning of SLL */
void insertion_at_begin()
{
	new1 = (struct link* ) malloc(sizeof(struct link));
	printf("\n Input the node value: ");
	scanf("%d", &new1->info);
	getchar();

	new1->next = start ;
	start = new1;
}

/* Function to insert a node at the end of SLL */
void insertion_at_end()
{
	struct link *temp=start;        /* Point to the first node of the list */

	new1 = (struct link* ) malloc(sizeof(struct link));

	printf("\n Input the node value: ");
	scanf("%d", &new1->info);
	getchar();

	if(start == NULL)
	{
		new1->next = start ;
		start = new1;
	}
	else
	{
		while (temp->next != NULL)  {   // To reach before the desired position
			temp = temp->next;
		}

		new1->next = temp->next ;
		temp->next = new1;
	}
}

/* Function to traverse the SLL */
void traverse()
{
	struct link *temp=start;      /* Point to the first node of the list */

	if(start == NULL)
		printf("\n Empty linked list");
	else
	{
		printf("\n List is as follows:\n");
		while (temp)
		{
			printf(" %d", temp->info);
			temp = temp->next;
		}
	}
}

void insertion_at_position() {
	int pos=0;
	int count=0;
	struct link *temp =start;
	
	while(temp) {
		temp=temp->next;
		count++;
	}

	if(count<2) {
		printf("there should be atleast 2 nodes to use insertion_at_choice");
	}else{
		printf("total node exist : %d\n",count);
		while(pos<2||pos>count) {
			printf("enter the position only betweem 2 and %d: ",count);
			scanf("%d",&pos);
		}
	    new1=(struct link*)malloc(sizeof(struct  link));
	    if(new1==NULL) {
		    printf("memory allocation failed!");
			exit(0);
		}
	    printf("enter value of new node : ");
	    scanf("%d",&new1->info);
	    getchar();

		temp=start;
		for(int i=1;i<pos-1;i++) {
			temp =temp->next;
		}
	    new1->next=temp->next;
	    temp->next=new1;
	}

}

void delete_by_position() {
	int pos=0;
	int count=0;
	struct link *temp=start;
	struct link *prev_temp;
	while(temp) {
		temp=temp->next;
		count++;
	}

	if(count==0) {
		printf("no node exist! first create one ");

	}else if(count==1) {
		free(start);
		printf("one node got deleted");

	}else {

		printf("total node : %d\n",count);

		do{
			printf("enter the position of node to be deleted(only between 1 and %d) : ",count);
		    scanf("%d",&pos);
		    getchar();
		}while(pos<1 || pos>count);

		temp=start;
		for(int i=1;i<pos;i++) {
			if(i==pos-1) {
				prev_temp=temp;
			}
			temp=temp->next;
		}

		if(pos==1) {
			start=temp->next;
			free(temp);
		}else {
			prev_temp->next=temp->next;
		    free(temp);
		}
		

	}    
}

void delete_by_value() {
	int value;
	int found=0;
	struct link *temp = start;
	struct link *prev_temp;
	if(temp==NULL) {
		printf("no node exist");
	}else {

		printf("enter the value of node to be deleted : ");
		scanf("%d",&value);
		getchar();

		if(value==temp->info) {        //for first node
		    found=1;
			start = temp->next;
			free(temp);
			printf("node deleted");

		}else if(temp->next!=NULL ){
			while(temp) {               //for other nodes
				prev_temp=temp;
				temp=temp->next;
				if(temp!=NULL && value==temp->info) {
					found=1;
					prev_temp->next=temp->next;
					free(temp);
					printf("node deleted");
					break;
				}
		    }
		}
		if(found ==0) {
			printf("value %d doesn't exist in list",value);
		}

		
	}
}

void search() {
	int value;
	struct link *temp=start;

	if(temp==NULL) {
		printf("no node exist");
	}else {
		printf("enter the value for searching : ");
		scanf("%d",&value);
		getchar();
		
		while(temp) {
		if(value==temp->info) {
			printf("ys value %d exist",value);
			break;
		}
		temp=temp->next;
		if(temp==NULL) {
			printf("value %d does not exist\n",value);
		}
	}
	}

}

 void reverse() {
	struct link *past=NULL;
	struct link *curr=start;
	struct link *next=NULL;

	if(curr==NULL) {
		printf("no linked list exist\n");
	}else {
		while(curr) {
			next = curr->next;
			curr->next=past;
			past=curr;
			curr=next;
		}
		start = past;
	}

	printf("reverse the list completed");
}

void short_ascending() {
	if(start== NULL) {
		printf("no linked list");
	}else{
		struct link *temp = start;
		struct link *temp1;
		int store;
		while(temp!=NULL) {
			temp1=  start;
			while(temp1!= NULL) {
				if(temp->info<temp1->info) {
					store = temp->info;
					temp->info= temp1->info;
					temp1->info= store;
				}
				temp1= temp1->next;
			}
			temp = temp->next;
		}
	}
}


//short linked list ascending or descending order

struct link *rcrs(struct link *temp) {
	if(temp == NULL || temp->next==NULL) {
		return temp;
	}
	struct link *newstart = rcrs(temp->next);
	temp->next->next=temp;
	temp->next=NULL;
	return newstart;
}

void shivam_reverse() {
	start = rcrs(start);
}
